components/raftstore/src/store/fsm/peer.rs

```diff
@@ -5311,6 +5378,17 @@ where
             replicated_idx
         };
         assert!(compact_idx >= first_idx);
+
+        if let Some(idx) = witness_replicated_idx {
+            // 1. do not compact the log not replicated to witness, except witness is the
+            // most lagging peer
+            // 2. if there is a lagging behind follower, should not compact log, otherwise
+            // witness can't help the follower to catch up if the leader is down
+            if idx != replicated_idx {
```

Comment:
This is too strict. You can't expect different nodes have the same replication speed. So they are likely to be different in a heavy write group.